### PR TITLE
Remove default empty state of `sf::Event`

### PR DESCRIPTION
**Pull request by vittorioromeo**
Proof of concept right now as I don't have access to a compiler, but I think this is a superior design compared to #2991. 

Removing the empty state from event results in a much more type-safe API. Events cannot be default-constructed anymore, and APIs that poll events now return an optional event, making it clear that an empty state can be present without polluting usages of events that cannot be in such an empty state.

@ChrisThrasher: feel free to continue working on this branch to adjust the usages of event in tests and examples.

### CODE REVIEW

**Comment by ChrisThrasher** [comment]
I considered this in the past and decided I didn’t like it because it requires excessive `operator->` usage. I find the current `operator bool` API more ergonomic even if it’s less ideologically pure.

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/compare/master...optional_events

Here's my branch from back in November. I looked at this and didn't think it was worth doing.

**Comment by L0laapk3** [comment]
What are the advantages of this over std::monostate?

**Comment by vittorioromeo** [comment]
> What are the advantages of this over std::monostate?

Monostate is just another symbol for ab empty state, which is a state a valid event should never be in.

As @ChrisThrasher mentioned in his PR, event shouldn't even have an empty state and most importantly such a state should not be public and accessible to users.

The reason why we had an empty state to begin with is because some APIa such as pollEvent need to return immediately and tell the user that no event was polled.

In C++17 we have an idiomatic way of expressing the possible absence of a result: optional. We are using that throughout the library already and using it for events solves the "public empty state but not really" issue and, even better, makes event a type that's always guaranteed to be valid and non empty.

**Comment by L0laapk3** [comment]
> ... since users already should not be using `sf::Event::Empty`.

@ChrisThrasher Can you explain the rationale behind this? What is wrong `sf::Event::Empty`?

**Comment by vittorioromeo** [comment]
> > ... since users already should not be using `sf::Event::Empty`.
> 
> @ChrisThrasher Can you explain the rationale behind this? What is wrong with `sf::Event::Empty`? Is it just about having to implement similar safeties to `std::monostate`?

It's a meaningless implementation detail. Operating systems or windowing systems do not produce empty events. We simply needed a sentinel to signal "hey, you tried to poll events but nothing was available" in user loops.

**Comment by vittorioromeo** [comment]
@eXpl0it3r @binary1248: could you please express your thoughts on this PR? It follows exactly the same philosophy as these already-merged PRs:

- #3020 
- #3021 
- #3022
- #3025
- #3028
- #3030
- #3032

And the same philosophy as the soon-to-be-merged PRs:

- #3027

In fact, I don't understand @ChrisThrasher's concerns given the fact that he's the author of all the PRs mentioned above.

It is true that the user experience gets a tiny bit worse due to the fact that the `->` token must be used compared to the `.` token, and due to the fact that the user needs to be familiar with `std::optional`, but I genuinely believe that is a blessing in disguise. 

Our users' will benefit from the type safety of `std::optional`, and once they learn the pattern of using it, they will appreciate the consistency of such pattern being used everywhere in the library, `sf::Event` included.

There are a few more benefits to this PR:

- We avoid exposing the `sf::Event::Empty` class publicly to the users, which was just an implementation detail;

- We avoid a public `operator bool` conversion on `sf::Event`, which is again an implementation detail exposed to the user, as we were pretty much reimplementing a `std::optional` on our own;

- `sf::Window::waitEvent` now guarantees to return a non-empty event, simplifying code that uses `waitEvent` instead of `pollEvent`;

- The signature of `sf::Window::pollEvent` now clearly shows that an event might not be returned, making it clear how it differs from `waitEvent` and alerting users that they need to check for the empty case;

- This PR opens the door to a very nice declarative visitation-based event processing technique, see #3015 and #3014.

In conclusion, I strongly believe that this PR helps us fulfill our goal of making invalid states unrepresentable in the type system. Not merging it would not only hinder us from reaching that goal, but would undermine the work done by @ChrisThrasher with his previously merged PRs due to a glaring inconsistency between most SFML types and `sf::Event`.

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 1]
```diff
@@ -44,8 +44,8 @@ body:
           label: Steps to reproduce
           description: Tell us how to reproduce this issue and provide a [minimal, complete and verifiable example](https://stackoverflow.com/help/mcve), you can use the template below
           value: |
-              1. 
-              2.
```
Comment (ChrisThrasher) [inline .github/ISSUE_TEMPLATE/bug_report.yml:48]: I think these trailing newlines may be intentional

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 2]
```diff
@@ -57,10 +57,13 @@ body:
 
                   while (window.isOpen())
                   {
-                      while (const auto event = window.pollEvent())
+                      while (const std::optional event = window.pollEvent())
```
Comment (ChrisThrasher) [inline .github/ISSUE_TEMPLATE/bug_report.yml:60]: As a point of style I'd prefer we continue to use `auto` here and in all other similar instances

**Comment by vittorioromeo** [comment]
> Can you elaborate on how this PR relates to supporting visitation?

I overstated my point, the same visitation API can be achieved even without this PR. The only difference is that without this PR, we'd have to inject a `sf::Event::Empty` overload ourselves to make the visitor passed to `std::visit` exhaustive.

This PR removes the need for such a "hack", sort of proving that everything fits better together.

Edited my original post.

File: .github/ISSUE_TEMPLATE/feature_request.yml [hunk 3]
```diff
@@ -54,10 +54,13 @@ body:
...
                   while (window.isOpen())
                   {
-                      while (const auto event = window.pollEvent())
+                      while (const std::optional event = window.pollEvent())
                       {
-                          if (event.is<sf::Event::Closed>())
+                          if (event->is<sf::Event::Closed>())
+                          {
                               window.close();
+                              break;
```
Comment (ChrisThrasher) [inline .github/ISSUE_TEMPLATE/feature_request.yml:62]: I suppose this `break` ensures that event processing ceases when a close is requested? In theory this reduces the latency between clicking the close button the application actually closing. Is that your intent?

File: include/SFML/Window/WindowBase.hpp [hunk 4]
```diff
@@ -180,35 +180,33 @@ class SFML_WINDOW_API WindowBase
...
     /// \code
-    /// while (const auto event = window.pollEvent())
+    /// while (const std::optional<sf::Event> event = window.pollEvent())
     /// {
     ///    // process event...
     /// }
     /// \endcode
     ///
-    /// \return The event; will be `Empty` (convertible to `false`) if no events are pending
+    /// \return The event; will be `std::nullopt` (convertible to `false`) if no events are pending
```
Comment (ChrisThrasher) [inline include/SFML/Window/WindowBase.hpp:194]:
```suggestion
    /// \return The event, otherwise `std::nullopt` if no events are pending
```
Tweaking the language to be more in line with how we document other `optional`-y APIs

File: src/SFML/Window/Android/WindowImplAndroid.cpp [hunk 5]
```diff
@@ -462,28 +462,20 @@ int WindowImplAndroid::processKeyEvent(AInputEvent* inputEvent, ActivityStates&
...
 int WindowImplAndroid::processMotionEvent(AInputEvent* inputEvent, ActivityStates& states)
 {
-    const std::int32_t device = AInputEvent_getSource(inputEvent);
-
-    Event event;
-
-    if (device == AINPUT_SOURCE_MOUSE)
-        event = Event::MouseMoved{};
-    else if (static_cast<std::uint32_t>(device) & AINPUT_SOURCE_TOUCHSCREEN)
-        event = Event::TouchMoved{};
```
Comment (ChrisThrasher) [inline src/SFML/Window/Android/WindowImplAndroid.cpp:472]: How is it that we can get rid of this?

File: src/SFML/Window/DRM/InputImpl.cpp [hunk 6]
```diff
@@ -353,9 +353,9 @@ bool eventProcess(sf::Event& event)
     static unsigned int doDeferredText = 0;
     if (doDeferredText)
     {
-        event          = sf::Event::TextEntered{doDeferredText};
+        auto event     = sf::Event::TextEntered{doDeferredText};
```
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/InputImpl.cpp:356]: Is this non-`const` to ensure a move can occur?

File: src/SFML/Window/WindowBase.cpp [hunk 7]
```diff
@@ -146,22 +146,29 @@ bool WindowBase::isOpen() const
...
 Event WindowBase::waitEvent()
 {
-    Event event;
-    if (m_impl && (event = m_impl->popEvent(true)))
-        filterEvent(event);
-    return event;
+    assert(m_impl);
+
+    std::optional<sf::Event> event = m_impl->popEvent(true /* blocking */);
+    assert(event.has_value());
```
Comment (ChrisThrasher) [inline src/SFML/Window/WindowBase.cpp:168]: Why `assert` here instead of keeping the original conditional? Can we safely assume `popEvent` never returns `std::nullopt`?

File: src/SFML/Window/WindowImpl.cpp [hunk 8]
```diff
@@ -201,14 +201,12 @@ Event WindowImpl::popEvent(bool block)
...
     // Pop the first event of the queue, if it is not empty
-    if (!m_events.empty())
-    {
-        event = m_events.front();
-        m_events.pop();
-    }
+    if (m_events.empty())
+        return std::nullopt;
+
+    sf::Event event = m_events.front();
```
Comment (ChrisThrasher) [inline src/SFML/Window/WindowImpl.cpp:208]:
```suggestion
    Event event = m_events.front();
```

File: src/SFML/Window/WindowImpl.hpp [hunk 9]
```diff
@@ -128,14 +128,14 @@ class WindowImpl
...
     /// if no event is available: if it is true then the function
     /// doesn't return until a new event is triggered; otherwise it
-    /// returns an empty event to indicate that no event is available.
+    /// returns `std::nullopt` to indicate that no event is available.
     ///
-    /// \param block Use true to block the thread until an event arrives
+    /// \param block Use `true` to block the thread until an event arrives
     ///
-    /// \return The event; can be `Empty` (convertible to `false`) if not blocking
+    /// \return The event; can be `std::nullopt` (convertible to `false`) if not blocking
```
Comment (ChrisThrasher) [inline src/SFML/Window/WindowImpl.hpp:135]:
```suggestion
    /// \return The event, otherwise `std::nullopt` if not blocking
```

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]
I suppose I can come around to this change. It is technically correct after all. My main argument about ergonomics is hard to support. Hopefully I'm wrong and users won't complain too much about using `->` so much in their event loops. 

Still curious to hear from the rest of the team.

**Comment by binary1248** [comment]
People can't simultaneously praise other hyped up languages for their at times really esoteric syntax and at the same time complain that C++ isn't staying up to date while complaining when ever so often we use already existing language constructs to make the language "less bad" than what its undeserved reputation has many laypeople believe.

Sure I also prefer having to type less, but not at the cost of higher mental load because I have to think of something that wouldn't have been necessary if I would have typed a bit more. I prefer typing over thinking if there was a choice.

I have used SFML for more than a decade at this point, and never have I spent more than 1% of total project time messing with event handling. If anyone spends so much time on event handling that this actually makes a difference they are doing something wrong anyway.

File: src/SFML/Window/WindowImpl.hpp (hunk 9 above)
Comment (vittorioromeo) [reply src/SFML/Window/WindowImpl.hpp:135]: This suggested change doesn't seem correct, it's not always `std::nullopt` if non blocking, but it *can be* `std::nullopt` only if it's non blocking

File: src/SFML/Window/WindowBase.cpp (hunk 7 above)
Comment (vittorioromeo) [reply src/SFML/Window/WindowBase.cpp:168]: Yes, `popEvent` in blocking mode will wait until an event is available. Will add a comment

Comment (vittorioromeo) [inline src/SFML/Window/WindowBase.cpp:168]:
```suggestion
    assert(event.has_value()); // `popEvent` in blocking mode guarantees to return an event
```

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 2 above)
Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/bug_report.yml:60]: We're basically doing that, only that we're constraining the type inference to a `std::optional` of some `T`. I think it's worthwhile using `std::optional` here instead of `auto` to clearly suggest that an event might or might not be polled.

I considered two alternatives:

- `auto`: hides too much information, not clear that a `std::optional` is returned;
- `std::optional<sf::Event>`: too verbose and redundant, the information about being an event is already clearly available from the name of the variable `event`.

Using `std::optional` strikes a good balance between conciseness and providing useful information to the reader.

File: examples/raw_input/RawInput.cpp [hunk 10]
```diff
@@ -36,7 +36,7 @@ int main()
     {
         while (const auto event = window.pollEvent())
```
Comment (vittorioromeo) [inline examples/raw_input/RawInput.cpp:37]:
```suggestion
        while (const std::optional event = window.pollEvent())
```

File: src/SFML/Window/WindowBase.cpp (hunk 7 above)
Comment (ChrisThrasher) [reply src/SFML/Window/WindowBase.cpp]: Is it overkill to refactor this into two functions where one returns `Event` and the other `std::optional<Event>`?

File: .github/ISSUE_TEMPLATE/feature_request.yml (hunk 3 above)
Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: That's not my intent, but it's a nice side benefit of the change.

The intention is to prevent event processing from causing issues when using `waitEvent`, as `waitEvent` now asserts `m_impl != nullptr`, but `WindowBase::close` immediately invokes `m_impl.reset()`.

We need to call `break;` to avoid accessing a deleted `m_impl` on the next iteration.

Now that I wrote this, it does sound like a subtle trap if the `break;` is forgotten... But the only way I can think of fixing this is to revert `waitEvent` to return a `std::optional<sf::Event>`, which loses a nice API improvement that this PR would provide.

I think a decent alternative would be to do this in `waitEvent`:

```cpp
Event WindowBase::waitEvent()
{
    if (m_impl == nullptr)
    {
        sf::err() << "waitEvent invoked on destroyed window -- did you forget a `break`?" << std::endl;
        std::terminate();
    }

    std::optional<sf::Event> event = m_impl->popEvent(true /* blocking */);
    assert(event.has_value());

    filterEvent(*event);
    return *event;
}
```

File: src/SFML/Window/Android/WindowImplAndroid.cpp (hunk 5 above)
Comment (vittorioromeo) [reply src/SFML/Window/Android/WindowImplAndroid.cpp:472]: That code did nothing, it's a no-op -- the same conditions are checked below when `forwardEvent` is called. I believe my change is equivalent in behavior to the previous code.

File: .github/ISSUE_TEMPLATE/feature_request.yml (hunk 3 above)
Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: I've added the additional safety measure to `waitEvent` and updated its documentation.

File: src/SFML/Window/WindowBase.cpp [hunk 11]
```diff
@@ -146,22 +147,36 @@ bool WindowBase::isOpen() const
...
-    if (m_impl && (event = m_impl->popEvent(true)))
-        filterEvent(event);
-    return event;
+    if (m_impl == nullptr)
+    {
+        err() << "Fatal error: `waitEvent` invoked on destroyed window -- did you forget a `break` after "
+                 "`window.close()` in your event handling loop?"
+              << std::endl;
+
+        std::abort();
```
Comment (vittorioromeo) [inline src/SFML/Window/WindowBase.cpp:172]: @ChrisThrasher: alternatively here we could just return another artificial `sf::Event::Closed`, but I'd rather promote the good practice of breaking after closing the window.

Comment (vittorioromeo) [reply src/SFML/Window/WindowBase.cpp:174]: @ChrisThrasher: I decided to do that in the end, it's much friendlier for the user compared to termination.

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 1 above)
Comment (eXpl0it3r) [reply .github/ISSUE_TEMPLATE/bug_report.yml:48]: Trailing spaces are intentional, so you start directly typing

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 2 above)
Comment (eXpl0it3r) [reply .github/ISSUE_TEMPLATE/bug_report.yml:60]: Before bringing up the same discussion point in every PR, maybe let's leave it as it is right now. Start a separate discussion and the align the code pieces afterwards.

It's not helpful when someone writes code in one way and someone else then goes and keeps changing it afterwards in their way.

File: .github/ISSUE_TEMPLATE/feature_request.yml (hunk 3 above)
Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: > We've found an edge case where `waitEvent` can return an empty state

It's more of a precondition failure -- the precondition is that `waitEvent` should not be called  on a window that's already been closed. Returning a `sf::Event::Closed` is a way of not terminating nor getting UB, but it's not behavior the user should rely on.

Perhaps I can add an `err()` before the return to make it clear that the user is doing something wrong and we're just being nice?

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 2 above)
Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/bug_report.yml:60]: That's fair, but there was a reason I made this change specifically in this PR, as it was now unobvious that `pollEvent` returns an optional with plain `auto'.

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 1 above)
Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/bug_report.yml:48]: Will fix when I get back home

File: .github/ISSUE_TEMPLATE/feature_request.yml (hunk 3 above)
Comment (eXpl0it3r) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: Actually, this just reminded me of the longstanding issue of `waitEvent` never returning if there are no more events to be produced. For example if you run `waitEvent` in a separate thread and in the other thread close the window. The event thread will continuing spinning as it doesn't get any events to process: https://github.com/orgs/SFML/projects/4?pane=issue&itemId=23482732# / https://github.com/SFML/SFML/issues/1184

Since SFML is anyways doing a spinning while loop, due to Joystick event processing, I'd actually propose to change `waitEvent` to take a custom timeout parameter. Blocking methods should always have a way to cancellation token or timeout option.

In that case you'll have another situation where waitEvent has a "failure case".

The user isn't doing anything wrong, when calling any of the window methods after closing the window. Because you still have a valid object and as such any operation on that object should remain valid for the lifetime of the object.

Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: > Actually, this just reminded me of the longstanding issue of `waitEvent` never returning if there no more events are being produced. For example if you run `waitEvent` in a separate thread and in the other thread closes the window. The event thread will continuing spinning as it doesn't get any events to process: https://github.com/orgs/SFML/projects/4?pane=issue&itemId=23482732# / #1184
> 
> Since SFML is anyways doing a spinning while loop, due to Joystick event processing, I'd actually propose to change `waitEvent` to take a custom timeout parameter. Blocking methods should always have a way a cancellation token or timeout option.
> 
> In that case you'll have another situation where waitEvent has a "failure case".
> 
> The user isn't doing anything wrong, when calling any of the window methods after closing the window. Because you still have a valid object and as such any operation on that object should remain valid for the lifetime of the object.

I like the timeout idea, I will work on a PR for that specifically (against master) and then integrate that into this one as they are separate concerns.

Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: Created PR for the timeout here: https://github.com/SFML/SFML/pull/3094

Comment (ChrisThrasher) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: > It's more of a precondition failure -- the precondition is that waitEvent should not be called on a window that's already been closed.

In my opinion calling `waitEvent` on a closed window is precondition violation. We should log an error, `assert(false)`, then return `sf::Event::Closed` to ensure behavior is defined in release builds. I still want that function to return an `sf::Event` rather than `std::optional<sf::Event>`.

Comment (eXpl0it3r) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: I strongly disagree. As long as an object was initialized, it's in a valid state. We cannot suddenly have functions semi-fail because we assign random preconditions.

Besides, with the timeout proposal, you actually get an expected failure case.

I'm also against adding `break;` everywhere.

Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: My views:

- It's very unfortunate that we lose the nice non-`optional` API for `waitEvent` if we add the timeout, but I think it's a good idea to have one;

- I agree with @ChrisThrasher that calling `waitEvent` on a closed window should be a precondition violation and I agree with his proposed solution -- however if we do add a timeout it would make sense to return `std::nullopt` there as well;

- I disagree with both of you regarding `break;` -- I think that the only sensible thing to do when you close a window is to break out of the event handling loop, and I'd like to promote that practice regardless of whether or not it's strictly necessary for the correctness of the program.

Comment (ChrisThrasher) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: I think adding `break` in all these places is unrelated to the core mission of this PR. If you want to make the case for it, I'd rather that happen in a separate PR because discussing it is stalling the more important API changes. In general it gets back to a personal preference of mine that PRs not try to solve multiple orthogonal problems at once.

As for `waitEvent` changes, does that PR need to be settled first before this PR's review can continue? Help us understand the best order of operations for all the PRs you have recently opened.

Comment (vittorioromeo) [reply .github/ISSUE_TEMPLATE/feature_request.yml:62]: #3094 was merged in, I've rebased this PR on that. I've also merged in #3096 into this PR, I would like to merge that as a prerequisite but I can extract it out if required. Resolving for now.

File: include/SFML/Window/Event.hpp [hunk 12]
```diff
@@ -275,12 +267,10 @@ class SFML_WINDOW_API Event
...
 
     ////////////////////////////////////////////////////////////
-    /// \brief Default constructor
-    ///
-    /// Sets the event to sf::Event::Empty
+    /// \brief Deleted default constructor
     ///
     ////////////////////////////////////////////////////////////
-    Event() = default;
+    Event() = delete;
```
Comment (ChrisThrasher) [inline include/SFML/Window/Event.hpp:273]:
```suggestion
```
This can be removed since it is already implicitly deleted. To ensure we don't regress in this respect we should add a corresponding test.

https://github.com/search?q=repo%3ASFML%2FSFML%20is_default_constructible_v&type=code

I added tests for `!std::is_default_constructible_v<sf::Foo>` for all types for which we removed their default constructor so making this change is in line with existing precedent.

File: include/SFML/Window/WindowBase.hpp [hunk 13]
```diff
@@ -210,20 +210,20 @@ class SFML_WINDOW_API WindowBase
     /// dedicated to events handling: you want to make this thread sleep
     /// as long as no new event is received.
     /// \code
-    /// if (const auto event = window.waitEvent())
+    /// while (const std::optional<sf::Event> event = window.waitEvent())
```
Comment (ChrisThrasher) [inline include/SFML/Window/WindowBase.hpp:213]:
```suggestion
    /// while (const std::optional event = window.waitEvent())
```

File: include/SFML/Window/WindowBase.hpp [hunk 14]
```diff
@@ -181,23 +181,23 @@ class SFML_WINDOW_API WindowBase
...
     ///
     /// This function is not blocking: if there's no pending event then
-    /// it will return an empty event. Note that more than one event
+    /// it will return a `std::nullopt`. Note that more than one event
     /// may be present in the event queue, thus you should always call
     /// this function in a loop to make sure that you process every
     /// pending event.
     /// \code
-    /// while (const auto event = window.pollEvent())
+    /// while (const std::optional<sf::Event> event = window.pollEvent())
```
Comment (ChrisThrasher) [inline include/SFML/Window/WindowBase.hpp:189]:
```suggestion
    /// while (const std::optional event = window.pollEvent())
```

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]
I think I'm ready to approve this if the `break` additions are removed. I'm still not convinced I want to add them and would rather leave them out of this PR in the meantime.

File: include/SFML/Window/Event.hpp (hunk 12 above)
Comment (vittorioromeo) [reply include/SFML/Window/Event.hpp:273]: Hmm, do you actually think it would be better to not have the explicit `= delete`? I think it conveys intent, and we do the same thing for some classes that cannot be copied or moved because of some of their members being non-copyable or non-movable. Technically the parent class doesn't need their copy/move constructors deleted, but we are still explicit about it.

I'd prefer leaving the explicit `= delete` in, but not a big deal for me if you believe it's best to rely on the implicit deletion.

Comment (ChrisThrasher) [reply include/SFML/Window/Event.hpp:273]: What I'd say is that the tests also convey intent but do so in a more terse manner. The type trait tests already cover many implicitly defined operations. Here's a good example of that: https://github.com/SFML/SFML/pull/2874

Comment (vittorioromeo) [reply include/SFML/Window/Event.hpp:273]: > What I'd say is that the tests also convey intent but do so in a more terse manner.

The tests do not appear in the public API or documentation of `sf::Event`. I still think it's worth keeping the deleted ctor, but I'd rather remove it and resolve this conversation and potentially re-introduce it later to avoid friction.

Comment (vittorioromeo) [reply include/SFML/Window/Event.hpp]: Removed the ctor and added a test

**Comment by vittorioromeo** [comment]
@ChrisThrasher: I have removed all the extra `break`s and rebased on `master`.

**Comment by vittorioromeo** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Window/WindowImpl.cpp [hunk 15]
```diff
@@ -212,13 +212,13 @@ Event WindowImpl::pollEvent()
...
-Event WindowImpl::popEvent()
+std::optional<Event> WindowImpl::popEvent()
 {
-    Event event;
+    std::optional<Event> event; // Use a single local variable for NRVO
 
     if (!m_events.empty())
     {
-        event = m_events.front();
+        event.emplace(m_events.front());
```
Comment (ChrisThrasher) [inline src/SFML/Window/WindowImpl.cpp:221]: What does `emplace` offer here over `operator=`?

File: src/SFML/Window/WindowBase.cpp [hunk 16]
```diff
@@ -146,22 +168,22 @@ bool WindowBase::isOpen() const
...
+std::optional<Event> WindowBase::pollEvent()
 {
-    Event event;
-    if (m_impl && (event = m_impl->pollEvent()))
-        filterEvent(event);
-    return event;
+    return WindowsBaseImpl::pollOrWaitEventImpl(
+        m_impl.get(),
+        [this] { return m_impl->pollEvent(); },
+        [this](sf::Event& e) { filterEvent(e); });
```
Comment (ChrisThrasher) [inline src/SFML/Window/WindowBase.cpp:176]: I personally think this is overabstraction, but won't demand its removal. I get how it works now but don't love how long it took to understand how it works compared to the naive alternative of having two functions with nearly identical implementations. I don't wish to see more of this pattern going forward.

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]
Thanks for the changes. I'm close to approving this pending a few minor comments. Does @eXpl0it3r have any final thoughts to share about this?

File: src/SFML/Window/WindowBase.cpp (hunk 16 above)
Comment (vittorioromeo) [reply src/SFML/Window/WindowBase.cpp:176]: Fair enough, the repetition is not that bad here, so I will remove the template function. I think this pattern is still valuable when there's a larger amount of repetition to ensure things are kept in sync and there are no copy-paste mistakes, hopefully you'll agree with that for future cases where it's more valuable.

File: include/SFML/Window/WindowBase.hpp [hunk 17]
```diff
@@ -210,20 +210,20 @@ class SFML_WINDOW_API WindowBase
     /// dedicated to events handling: you want to make this thread sleep
     /// as long as no new event is received.
     /// \code
-    /// if (const auto event = window.waitEvent())
+    /// while (const std::optional event = window.waitEvent())
```
[the remark this answers is not in the document]
Comment (vittorioromeo) [reply include/SFML/Window/WindowBase.hpp:213]: Will fix

File: src/SFML/Window/WindowImpl.cpp (hunk 15 above)
Comment (vittorioromeo) [reply src/SFML/Window/WindowImpl.cpp:221]: It makes it clear that `m_events.front()` returns an `sf::Event` and not a `std::optional<sf::Event>`. If I used `operator=`, it wouldn't be clear from this line of code whether `m_event` was a container of events or of optional events. `.emplace` makes it clear.

**Review by ChrisThrasher — APPROVED** [review]
I'd still like to get Exploiter's approval but this LGTM. Thanks for your patience over all the rounds of review and extensive discussion. This should be the last `std::optional` API change which means we'll have officially completed all our v3 API goals! I'm very excited as this marks a huge step towards freezing the v3 API in preparation of shipping.

File: src/SFML/Window/DRM/InputImpl.cpp [hunk 18]
```diff
@@ -510,7 +503,7 @@ bool eventProcess(sf::Event& event)
     if (ready > 0 && FD_ISSET(STDIN_FILENO, &readFDSet))
         bytesRead = read(STDIN_FILENO, &code, 1);
 
-    if ((code == 127) || (code == 8)) // Suppress 127 (DEL) to 8 (BACKSPACE)
+    if ((code == 127) || (code == 8)) // Suppress127 (DEL) to 8 (BACKSPACE)
```
Comment (kimci86) [inline src/SFML/Window/DRM/InputImpl.cpp:506]: Unintended space removal here

File: src/SFML/Window/DRM/InputImpl.cpp [hunk 19]
```diff
@@ -603,7 +592,7 @@ String getDescription(Keyboard::Scancode /* code */)
 }
 
 
-////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////
```
Comment (kimci86) [inline src/SFML/Window/DRM/InputImpl.cpp:595]: Looks like an unintended change too.

File: src/SFML/Window/WindowBase.cpp [hunk 20]
```diff
@@ -43,6 +43,7 @@
 #include <vector>
 
 #include <cassert>
+#include <cstdlib>
```
Comment (kimci86) [inline src/SFML/Window/WindowBase.cpp:46]: Why?

**Review by kimci86 — COMMENTED** [review]
Too late but here are some minor comments.

File: src/SFML/Window/DRM/InputImpl.cpp (hunk 18 above)
Comment (ChrisThrasher) [reply src/SFML/Window/DRM/InputImpl.cpp:506]: https://github.com/SFML/SFML/pull/3127
